File: py/day_02/day_02.py
#!/usr/bin/env python3

import os

values = {
    "A": 1, # A, X = "Rock" # 1
    "X": 1,
    "B": 2, # B, Y = "Paper" # 2
    "Y": 2,
    "C": 3, # C, Z = "Scissors" # 3
    "Z": 3
}

def rps_result(opponent: str, my_choice: str) -> str:

    if (opponent == "A" and my_choice == "Y") or \
        (opponent == "B" and my_choice == "Z") or \
        (opponent == "C" and my_choice == "X"):
        return "w"

    if (opponent == "A" and my_choice == "Z") or \
        (opponent == "B" and my_choice == "X") or \
        (opponent == "C" and my_choice == "Y"):
        return "l"

    else:
        return "d"


def rps_reverse(opponent: str, outc: str) -> str:

    possibilities = ["X", "Y", "Z"]

    if outc == "Y": #draw
        return opponent

    if outc == "Z": #win
        for pos in possibilities:
            if rps_result(opponent, pos) == 'w':
                return pos

    else: # lose
        for pos in possibilities:
            if rps_result(opponent, pos) == 'l':
                return pos



def solve_part1(round_results: list) -> int:

    total_score = 0
    for janken_round in round_results:
        opp, my_choice = janken_round.split()

        choice_val = values[my_choice]

        res = rps_result(opp, my_choice)
        if res == "w":
            match_val = 6
        elif res == "d":
            match_val = 3
        else:
            match_val = 0

        total_score += choice_val + match_val


    return total_score

def solve_part2(round_results: list) -> int:
    # X = lose
    # Y = draw
    # Z = win

    total_score = 0
    for janken_round in round_results:
        opp, outc = janken_round.split()

        if outc == "Z":
            match_val = 6
        elif outc == "Y":
            match_val = 3
        else:
            match_val = 0

        needed_shape = rps_reverse(opp, outc)

        choice_val = values[needed_shape]

        total_score += choice_val + match_val

    return total_score

if __name__ == "__main__":
    directory = os.path.abspath(os.path.dirname(__file__))
    with open(os.path.join(directory, "input.txt"), encoding="utf-8") as file:
        problem_set = file.read().splitlines()
    print(solve_part1(problem_set))
    print(solve_part2(problem_set))

File: kt/src/day02/Day02.kt
package day02

import readInput

fun winConditionFulfilled(opp: String, mine: String): Boolean{
    return (opp == "A" && mine == "Y") ||
            (opp == "B" && mine == "Z") ||
            (opp == "C" && mine == "X")
}

fun loseConditionFulfilled(opp: String, mine: String): Boolean{
    return (opp == "A" && mine == "Z") ||
            (opp == "B" && mine == "X") ||
            (opp == "C" && mine == "Y")
}

fun rpsResult(opponent: String, mine: String): String {

    if (winConditionFulfilled(opponent, mine)) {
        return "w"
    }

    if (loseConditionFulfilled(opponent, mine)) {
        return "l"
    }

    return "d"
}

fun rpsReverse(opp: String, outc: String): String {
    val possibilities = listOf("X", "Y", "Z")

    // draw
    if (outc == "Y") return opp

    // win
    if (outc == "Z") {
        for (out in possibilities) {
            if (rpsResult(opp, out) == "w"){
                return out
            }
        }
    }

    // lose
    for (out in possibilities) {
        if (rpsResult(opp, out) == "l"){
            return out
        }
    }

    return "error"
}
fun main() {

    val values = mapOf(
        "A" to 1, // A, X = "Rock" # 1
        "X" to 1,
        "B" to 2, // B, Y = "Paper" # 2
        "Y" to 2,
        "C" to 3, // C, Z = "Scissors" # 3
        "Z" to 3
    )

    fun part1(input: List<String>): Int {

        var totalScore = 0
        for (jankenRound in input) {
            val (opp, mine) = jankenRound.split(" ")

            val chosenShapeVal = values[mine]

            val matchVal =
                when (rpsResult(opp, mine)) {
                    "w" -> 6
                    "d" -> 3
                    else -> 0
                }

            totalScore += chosenShapeVal?.plus(matchVal) ?: 0
        }

        return totalScore
    }

    fun part2(input: List<String>): Int {

        var totalScore = 0
        for (jankenRound in input) {
            val (opp, outc) = jankenRound.split(" ")

            val matchVal =
                when (outc) {
                    "Z" -> 6
                    "Y" -> 3
                    else -> 0
                }

            val neededShape = rpsReverse(opp, outc)

            val chosenShapeVal = values[neededShape]

            totalScore += chosenShapeVal?.plus(matchVal) ?: 0
        }

        return totalScore
    }

    // test if implementation meets criteria from the description, like:
    val testInput = readInput("day02/test_input")
    check(part1(testInput) == 15)

    val input = readInput("day02/input")
    println(part1(input))
    println(part2(input))
}
